IPIP/0000-ordered-car-responses.md

```diff
@@ -0,0 +1,236 @@
+1. Blocks can be reused multiple times, this is handy for cases when you plan to cache on disk but not at all when you want to process a stream with use & forget policy.
+
+What we really want is for the gateway to help us a bit, and give us blocks in a useful order.
+
+## Detailed design
+
+Add a qualifying parameter to the `Accept` Header to how the blocks should be transfered.
+
+The block order correspond to someone reading the car file `Data` section (see car specification) from low index to high indexes.
+In other words, blocks are sent first when streaming the car file over a fifo byte stream such as TCP connection, unix pipe, ... they are readable first.
+
+`BlockOrder`, `enum (string)`, defaults to `Unordered`
+
+They are generally adding increasing level of compatible constraints,
+this mean a server implementation might choose to implement only some of the most restrictive one and use overly restrictive one for continence.
+The Compatibility graph looks like this:
+```graphviz
+digraph G {
+  {ForwardDFSWithSkippingOfDuplicates, BackwardDFSWithSkippingOfDuplicates} -> ProvenUsefullFirst
+  ForwardDFSWithDuplicates -> ForwardDFSWithSkippingOfDuplicates
+  BackwardDFSWithDuplicates -> BackwardDFSWithSkippingOfDuplicates
+
+  "ProvenUsefullFirst-OnlyUsefull" -> {ProvenUsefullFirst, "Unordered-OnlyUsefull"} -> Unordered
+  "ForwardDFSWithSkippingOfDuplicates-OnlyUsefull" -> {ForwardDFSWithSkippingOfDuplicates, "ProvenUsefullFirst-OnlyUsefull"}
+  "BackwardDFSWithSkippingOfDuplicates-OnlyUsefull" -> {BackwardDFSWithSkippingOfDuplicates, "ProvenUsefullFirst-OnlyUsefull"}
+  "ForwardDFSWithDuplicates-OnlyUsefull" -> ForwardDFSWithDuplicates
+  "BackwardDFSWithDuplicates-OnlyUsefull" -> BackwardDFSWithDuplicates
+
+  Query
+}
+```
+
+Compatible means that gateway is MAY answer with childs (direct or indirect) of a particular ordering if it wish to.
+
+A gateway MUST answer with a stronger compatible ordering if one is available.
+
+A gateway is allowed and SHOULD answer with a more strict `BlockOrder` type if this is what is really used under the hood.
+For example an implementation that doesn't do any particular more optimisation when asked `Unordered` and actually answer `ProvenUsefullFirst` SHOULD set `Content-Type` to `application/vnd.ipld.car; BlockOrder=ProvenUsefullFirst`.
+
+### Orderings
+
+Note for all of them, we follow multihash keying,
+if they allows not sending the same blocks multiple times, if multihash A has been sent with codec 1,
+and then if multihash A is seen again but with codec 2, you may not send this block,
+however you MUST NOT stop the traversal because the same hash decoded two way may have different descent.
+
+For gateways:
+
+`Unordered` is the only MUST implement ordering.
+
+- `ProvenUsefullFirst`
+- `ForwardDFSWithSkippingOfDuplicates`
+- `ForwardDFSWithDuplicates`
+
+are SHOULD implement orderings.
+
+- `BackwardDFSWithSkippingOfDuplicates`
+- `BackwardDFSWithDuplicates`
+
+are MAY implement orderings.
```

Comment:
I see 5 types of ordering, but no rationale why we need them.

Which one will be used by a low memory IoT device which is fetching firmware update as a CAR?
Are there other use cases that require different orders and handling of duplicates?